pkg/controllers/providers/instancetype/vmcapacitycache/controller.go

```diff
@@ -0,0 +1,105 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vmcapacitycache
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instancetype"
+	"github.com/awslabs/operatorpkg/reasonable"
+	"github.com/awslabs/operatorpkg/singleton"
+	"go.uber.org/multierr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+)
+
+type Controller struct {
+	kubeClient           client.Client
+	instancetypeProvider *instancetype.DefaultProvider
+	initialized          bool
+}
+
+func NewController(kubeClient client.Client, instancetypeProvider *instancetype.DefaultProvider) *Controller {
+	return &Controller{
+		kubeClient:           kubeClient,
+		instancetypeProvider: instancetypeProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, r reconcile.Request) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "instancetypes.vmcapacitycache")
+	if !c.initialized {
+		return c.reconcileAllNodes(ctx)
+	}
+	if err := c.instancetypeProvider.UpdateVMCapacityCache(ctx, c.kubeClient, r.Name); err != nil {
+		return reconcile.Result{}, fmt.Errorf("updating vm capacity cache, %w", err)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) reconcileAllNodes(ctx context.Context) (reconcile.Result, error) {
```

Comment:
With the changes to the registration function I suggested in a separate comment, we can eliminate this case altogether.